File: src/2141.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

long long maxRunTime(int n, int* batteries, int batteriesSize) {
    long long retVal = 0;

    int i;

    long sumPower = 0;
    for (i = 0; i < batteriesSize; ++i) {
        sumPower += batteries[i];
    }

    long long middle;
    long long left = 1;
    long long right = sumPower / n;
    while (left < right) {
        middle = right - (right - left) / 2;

        long long total = 0;
        for (i = 0; i < batteriesSize; ++i) {
            total += fmin(batteries[i], middle);
        }
        if (total >= (long long)(n * middle)) {
            left = middle;
        } else {
            right = middle - 1;
        }
    }
    retVal = left;

    return retVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (int)(1e5)
    struct testCaseType {
        int n;
        int batteries[MAX_SIZE];
        int batteriesSize;
    } testCase[] = {{2, {3, 3, 3}, 3}, {2, {1, 1, 1, 1}, 4}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: n = 2, batteries = [3,3,3]
     *  Output: 4
     *
     *  Input: n = 2, batteries = [1,1,1,1]
     *  Output: 2
     */

    long long answer;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: n = %d, dist = [", testCase[i].n);
        for (j = 0; j < testCase[i].batteriesSize; ++j) {
            printf("%s%d", (j == 0) ? "" : ",", testCase[i].batteries[j]);
        }
        printf("]\n");

        answer = maxRunTime(testCase[i].n, testCase[i].batteries, testCase[i].batteriesSize);
        printf("Output: %lld\n", answer);

        printf("\n");
    }

    return EXIT_SUCCESS;
}

File: src/2141.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def maxRunTime(self, n: int, batteries: List[int]) -> int:
        retVal = 0

        left = 1
        right = sum(batteries) // n
        while left < right:
            middle = right - (right - left) // 2

            total = 0
            for power in batteries:
                total += min(power, middle)
            if total // n >= middle:
                left = middle
            else:
                right = middle - 1

        retVal = left

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for n, batteries in zip([2, 2], [[3, 3, 3], [1, 1, 1, 1]]):
            # /* Example
            #  *  Input: n = 2, batteries = [3,3,3]
            #  *  Output: 4
            #  *
            #  *  Input: n = 2, batteries = [1,1,1,1]
            #  *  Output: 2
            #  */
            logging.info("Input: n = %s, batteries = %s", n, batteries)

            retVal = pSolution.maxRunTime(n, batteries)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)
